webapp/channels/src/packages/mattermost-redux/src/actions/profiles_statuses_groups.ts

```diff
@@ -0,0 +1,170 @@
+import type {ActionFunc} from 'mattermost-redux/types/actions';
+
+const pendingUserIdsForProfiles = new Set<string>();
+const MAX_USER_PROFILES_BATCH = 100;
+const USER_PROFILES_DURATION = 5 * 1000;
+let userProfilesIntervalId: NodeJS.Timeout | null = null;
+
+const pendingUserIdsForStatuses = new Set<string>();
+const MAX_USER_STATUSES_BUFFER = 200;
+const USER_STATUSES_DURATION = 5 * 1000;
+const USER_STATUSES_REQUEST_SURGE_THRESHOLD = 1000;
+let userStatusesIntervalId: NodeJS.Timeout | null = null;
+let haveUserStatusRequestsSurged = false;
+
+function addUserIdForProfiles(userId: string): ActionFunc<boolean> {
+    return (dispatch) => {
+        function getPendingProfilesById() {
+            console.log('addUserIdForProfiles start', pendingUserIdsForProfiles);
+
+            dispatch(getProfilesByIds(Array.from(pendingUserIdsForProfiles)));
+            pendingUserIdsForProfiles.clear();
+        }
+
+        // Process immediately if the pending user ids exceeds the limit
+        if (pendingUserIdsForProfiles.size >= MAX_USER_PROFILES_BATCH) {
+            console.log('addUserIdForProfiles', 'executing immediately for', pendingUserIdsForProfiles.size);
+
+            getPendingProfilesById();
+        }
+
+        pendingUserIdsForProfiles.add(userId);
+
+        // Start the interval if it is not already running
+        if (userProfilesIntervalId === null) {
+            userProfilesIntervalId = setInterval(() => {
+                if (pendingUserIdsForProfiles.size > 0) {
+                    console.log('addUserIdForProfiles', 'executing with interval for pending', pendingUserIdsForProfiles.size);
+
+                    getPendingProfilesById();
+                } else {
+                    console.log('addUserIdForProfiles', 'no pending user ids');
+                }
+            }, USER_PROFILES_DURATION);
+        }
+
+        return {data: true};
+    };
+}
+
+export function cleanupUserProfilesInterval() {
+    if (userProfilesIntervalId !== null) {
+        clearInterval(userProfilesIntervalId);
+        userProfilesIntervalId = null;
+    }
+}
+
+function addUserIdForStatuses(userId: string): ActionFunc<boolean> {
+    return (dispatch) => {
+        function getPendingStatusesById() {
+            console.log('addUserIdForStatuses start', pendingUserIdsForStatuses);
+
+            // Since we can only fetch 100 user statuses at a time, we need to batch the requests
+            if (pendingUserIdsForStatuses.size >= MAX_USER_STATUSES_BUFFER) {
+                // We use temp buffer here to store up until max buffer size
+                // and clear out processed user ids
+                const userIds: string[] = [];
+
+                let counter = 0;
+                for (const pendingUserId of pendingUserIdsForStatuses) {
+                    userIds.push(pendingUserId);
+                    pendingUserIdsForStatuses.delete(pendingUserId);
```

Comment:
I'm trying to think if there's a better way to take 100 items from the Set at once. I guess this is probably performs at least as well as if `pendingUserIdsForStatuses` was an array that we could just take the first 100 items of

Reply:
There isnt none right now, both Set and Array in this case as Data structure have some or the other trade offs